linode/data_source_linode_firewall.go

```diff
@@ -0,0 +1,189 @@
+package linode
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/linode/linodego"
+)
+
+func dataSourceLinodeFirewallRule() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"label": {
+				Type:        schema.TypeString,
+				Description: `Used to identify this rule. For display purposes only.`,
+				Computed:    true,
+			},
+			"action": {
+				Type:        schema.TypeString,
+				Description: `Controls whether traffic is accepted or dropped by this rule. Overrides the Firewall’s inbound_policy if this is an inbound rule, or the outbound_policy if this is an outbound rule.`,
+				Computed:    true,
+			},
+			"ports": {
+				Type:        schema.TypeString,
+				Description: `A string representation of ports and/or port ranges (i.e. "443" or "80-90, 91").`,
+				Computed:    true,
+			},
+			"protocol": {
+				Type:        schema.TypeString,
+				Description: "The network protocol this rule controls.",
+				Computed:    true,
+			},
+			"ipv4": {
+				Type: schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: "A list of IP addresses, CIDR blocks, or 0.0.0.0/0 (to allow all) this rule applies to.",
+				Computed:    true,
+			},
+			"ipv6": {
+				Type: schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: "A list of IPv6 addresses or networks this rule applies to.",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func dataSourceLinodeFirewallDevice() *schema.Resource {
```

Comment:
maybe it would make sense to make this function in `resource_linode_firewall.go` so it can be used in both the datasource and resource, as both are fully computed schemas.

Reply:
Great idea!